### PR TITLE
completed the 8-10 java exercises

### PR DESCRIPTION
**Pull request by mikeippi123**
I have completed the java exercises 8 - 10 and have put them all in sperate folders that are labeled exercise8, exercise9, and exercise10. I have also attached out put images labeled java_exercise8_output, java_exercise9_output, java_exercise10_output.
(Updated) resolved all comments left on exercises 8-10 and pushed updated versions of that code. Also pushed exercise 11 and exercise 1 output. the new files attached for exercise 11 are ShoppingCart1.java, ShoppingCartTest.java, and Product1.java. the output file is java_exercise11_output.PNG

(Second Update) I have fixed all the issues with exercise 11 as well as the test file and got correct outputs. I also created a check for null or empty hashmaps for exercise 8. I also attached an updated output for exercise 11 with all the tests. New file added was ShoppingCarDemo.java with main output calling classes from ShoppingCart1.java, Product1.java.

java_exercise11_output

java_exercise9_output
java_exercise10_output
java_exercise8_output

### CODE REVIEW

File: michaelippolito1/java/exercise9/FileReaderDemo.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+
+public class FileReaderDemo {
+
+    public static void main(String[] args) {
+        try (Scanner s = new Scanner(new File("C:\\Users\\QX533SG\\Documents\\Week2\\cdp\\michaelippolito1\\java\\exercise9\\words.txt"))) {
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise9/FileReaderDemo.java:14]: Allowing input of the file path would be nice.

File: michaelippolito1/java/exercise9/FileReaderDemo.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+           
+                   
+              
+
+
+            System.out.println();
+            System.out.println();
+            System.out.println("Words displayed in reverse: ");
+           
+            revlist(plural);
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise9/FileReaderDemo.java:70]: A reversed index loop would work well here: for (int i = plural.size() -1; i >= 0; i--).  But what you have works.

File: michaelippolito1/java/exercise9/FileReaderDemo.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+
+
+
+            System.out.println();
+            System.out.println();
+            System.out.println("Plural words removed: ");
+            for (int i = 0; i < list.size(); i++) {
+                
+                if (list.get(i).endsWith("s") == false && list.get(i).endsWith("s.") == false) {
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise9/FileReaderDemo.java:83]: In this check, you are missing the exceptions for "This" and "is".
If you created an "isPlural" method that checked the word, then that logic would be in 1 place.

**Review by rbongiorno-ey — CHANGES_REQUESTED** [review]

File: michaelippolito1/java/exercise11/Product1.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package michaelippolito1.java.exercise11;
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/Product1.java:1]: In ShoppingCart1, you redeclare this class as an internal class.
You should import Product1 in the ShoppingCart1 class file, and use this definition.

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
+package michaelippolito1.java.exercise11;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+
+public class ShoppingCart1 {
+
+    public static class Product1 {
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:10]: Import class definition from Product1.java

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            System.out.println("This item is not currently in the cart");
+
+        } else {
+            this.sum -= item3.itemPrice;
+        }
+
+        return takeOut;
+    }
+
+  public static void main(String[] args) {
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:71]: You should create a ShoppingCartDemo class that would have the main method, so that ShoppingCart1 stands as a real Object (the Shopping Cart), separate from the application (Shopping Cart Demo)

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    Scanner scnr = new Scanner(System.in);
+    Scanner scnr1 = new Scanner(System.in);
+    Scanner scnr2 = new Scanner(System.in);
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:74]: You only need 1 scanner for System.in.

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    
+    System.out.println(newCart.getNumberOfItems());
+    //Item 1
+    System.out.println("Enter the item name: ");
+    productName = scnr.nextLine();
+
+    System.out.println("Enter the item price: ");
+    productPrice = scnr.nextDouble();
+
+    Product1 item = new Product1(productName, productPrice);
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:88]: You can declare _Product item_ once.  Then with each product input, you just reuse the item variable, instead of item, item1, item2.

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    newCart.addItem(item);
+    
+    //Item 2
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:91]: For a repetitive process, use a loop.
for (int i = 0; i < 3; i++) { prompt for and add item }

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    System.out.println("Enter the item price: ");
+    productPrice = scnr2.nextDouble();
+    
+    Product1 item3 = new Product1(productName, productPrice);
+    newCart.addItem(item3);
+
+  
+
+    //prints total
+    System.out.println(newCart.getTotal());
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:114]: Add some context to your output.  "Cart total: "+newCart.getTotal()

File: michaelippolito1/java/exercise11/ShoppingCart1.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+    //prints new total with item 3 removed
+    System.out.println(newCart.getTotal());
+
+    scnr.close();
+    scnr1.close();
+    scnr2.close();
+  return;
+   }
+
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCart1.java:127]: Implementations are missing.
Are these intended to be overloaded methods, so you can either add/remove a Product object or add/remove using the product name & price?

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        assertEquals(expected, actual);
+    }
+
+    @Test // When empty, the cart has 0 items
+    void displayWhenCartisEmptiesd_Scenario1() {
+        // given
+        int expected = 0;
+        cart = new ShoppingCart1();
+        this.cart = new ShoppingCart1();
+        Product1 product1 = new Product1("Milk", 2.99);
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:35]: You don't actually use product1 because you are using the methods that take the name and price.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        // when
+        this.cart.addItem("Milk", 2.99);
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:37]: If you added an assertion after this step, to make sure the "Milk" was added, you would see that this method did not actually work.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        this.cart.removeItem("Milk", 2.99);
+        
+        // then
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:40]: This test incorrectly passed, because the cart starts with 0 items. The methods you called did not actually do any processing.  So the cart still had 0 items.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    @Test // When a new product is added, the number of items must be incremented
+    void displayWhenNewProductAdded() {
+        //given
+        int expected = 0;
+        cart = new ShoppingCart1();
+        this.cart = new ShoppingCart1();
+        
+        //when
+        this.cart.addItem("Milk", 2.99);
+        //then
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:53]: You expected value is incorrect.  Should have been 1.
As mentioned, addItem(name, price) does not actually do anything, so this test should have failed.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            // When a new product is added, the new balance must be the sum of the previous
+     @Test  // balance plus the cost of the new product
+    void displayNewBalanceWhenNewProductisAdded() {
+         //given
+         int expected = 0;
+         cart = new ShoppingCart1();
+         this.cart = new ShoppingCart1();
+         //when
+         this.cart.addItem("Milk", 2.99);
+         //then
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:66]: Same comments as above.  expected is incorrect than since addItem doesn't do anything, it did not fail.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+       int expected = 0;
+       cart = new ShoppingCart1();
+       this.cart = new ShoppingCart1();
+       //when
+       this.cart.addItem("Milk", 2.99);
+       this.cart.addItem("Eggs", 2.99);
+       this.cart.addItem("Bread", 2.99);
+       this.cart.removeItem("Milk", 2.99);
+
+       //then
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:82]: Same issue.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Test // When a product not in the cart is removed, verify if the proper error message
+          // is thrown
+    void displayErrorWhenProductisNotFound() {
+          //given
+          int expected = 0;
+          cart = new ShoppingCart1();
+          this.cart = new ShoppingCart1();
+          //when
+          this.cart.removeItem("Milk", 2.99); // this will thorw a error message because a new list was created and there are no items called "Milk"
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:95]: Nothing happened here.

File: michaelippolito1/java/exercise11/ShoppingCartTest.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+   
+          //then
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise11/ShoppingCartTest.java:97]: to Assert that removeItem failed, you would need a return from removeItem indicating success/failure, and then assert that it failed.

File: michaelippolito1/java/exercise8/MapDemo.java [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+		hashmap.put("Stuart", "Reges");
+		hashmap.put("Jessica", "Miller");
+		hashmap.put("Amanda", "Camp");
+		hashmap.put("Hal", "Perkins");
+
+        OutputResult(hashmap1);
+        OutputResult(hashmap2);
+        OutputResult(hashmap);
+
+
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise8/MapDemo.java:64]: Needs test for null and empty (new) hashmap.

**Review by rbongiorno-ey — CHANGES_REQUESTED** [review]

File: michaelippolito1/java/exercise8/MapDemo.java [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+        // Map is true due to no duplicates
+        hashmap.put("Marty", "Stepp");
+        hashmap.put("Stuart", "Reges");
+        hashmap.put("Jessica", "Miller");
+        hashmap.put("Amanda", "Camp");
+        hashmap.put("Hal", "Perkins");
+
+        // checks to see if hashmap is null or empty
+        if (hashmap != null && hashmap.isEmpty() != true) {
```
Comment (rbongiorno-ey) [inline michaelippolito1/java/exercise8/MapDemo.java:57]: You can put the check for null/empty and 'else' output inside the OutputResult method, that way you only have to code it once.

**Review by rbongiorno-ey — APPROVED** [review]
1 comment on how it could be improved. but all working now.
